### PR TITLE
watchtower/wtclient: reliable, asynchronous pipeline for revoked state backups

### PR DESCRIPTION
**Pull request by cfromknecht**
This PR presents the bulk of the `wtclient` package, a reliable subsystem that backs up revoked states to a configured watchtower. #2516 introduced the logic for constructing justice transactions and encrypting them. Here we add the remainder of the capabilities that aggregates all revoked states, assigns them to a particular session, then signs justice transactions and uploads them to the proper tower.

Great care is taken to ensure that the `wtclient.TowerClient` can do all of this in a non-blocking fashion, ensuring that we maintain LND's high throughput and performance in HTLC forwarding. However, this too creates difficulties of its own, as we must ensure that the whole process is reliable, and will continue until completion if the daemon is shutdown or killed intermittently. Thus, the client is implemented as an asynchronous pipeline to which breached states are added, and the client makes its best effort to upload them before allowing the daemon to exit.

Additional considerations are also given to mobile, which is intended to be the predominant platform in which the client is run and where it adds the most benefit. Since transient factors such as network connectivity may impede a successful upload during shutdown, the entire pipline can be `ForceQuit` to exit promptly. On OS's like iOS, where applications only have a short time to live once a shutdown is requested, this ensures that we terminate gracefully, and can pick back up where we left off once the daemon has been restarted.

The initial rollout will only be targeted at supporting a single, private tower. However, much of the logic in this package is included for more advanced features such as rotating between several target towers during session negotiation, or allowing more asynchrony between session negotiation which can be useful for added privacy. These can be enabled down the road once more testing is in place as we continue to improve on minor reliability edge cases in the single tower case.

The unit tests in this area in the process of being refactored, to enable them to be more generalized and table-driven. These will be posted later this week.

There's a lot more detail I can go into about the design, but it's getting late here in SF so I'll pick this up tomorrow 🚀

Depends on:
 - [x] #2604 
 - [x] #2605 
 - [x] #2606 
 - [x] #2615

### CODE REVIEW

File: watchtower/wtclient/client.go [hunk 1]
```diff
@@ -0,0 +1,722 @@
...
+
+	// Policy is the session policy the client will propose when creating
+	// new sessions with the tower. If the policy differs from any active
+	// sessions recorded in the database, those sessions will be ignored and
+	// new sessions will be requested immediately.
+	Policy wtpolicy.Policy
+
+	// PrivateTower is the net address of a private tower. The client will
+	// try to create all sessions with this tower.
+	PrivateTower *lnwire.NetAddress
```
Comment (Roasbeef) [inline watchtower/wtclient/client.go]: Not a blocker as is, but this should be updated to allow it to point to multiple towers. Additionally (for our use case elsewhere), we should be able to dynamically add _new_ towers to an existing running client. Based on my current understanding, we should just be able to add the new tower addr to the existing round robin iterator.

File: watchtower/wtclient/client.go [hunk 2]
```diff
@@ -0,0 +1,722 @@
...
+		// 4. Since all valid tasks have been assigned to session
+		// queues, we no longer need to negotiate sessions.
+		c.negotiator.Stop()
+
+		log.Debugf("Waiting for active session queues to finish "+
+			"draining, stats: %s", c.stats)
+
+		// 5. Shutdown all active session queues in parallel. These will
+		// exit once all updates have been acked by the watchtower.
+		c.activeSessions.ApplyAndWait(func(s *sessionQueue) func() {
```
Comment (Roasbeef) [inline watchtower/wtclient/client.go]: Scratch my earlier comment w.r.t usage of this method ;)

File: watchtower/wtclient/client.go [hunk 3]
```diff
@@ -0,0 +1,722 @@
...
+		log.Infof("Watchtower client unclean shutdown complete, "+
+			"stats: %s", c.stats)
+	})
+}
+
+// RegisterChannel persistently initializes any channel-dependent parameters
+// within the client. This should be called during link startup to ensure that
+// the client is able to support the link during operation.
+func (c *TowerClient) RegisterChannel(chanID lnwire.ChannelID) error {
+	// TODO(conner): create sweep pkscript for new channels
```
Comment (Roasbeef) [inline watchtower/wtclient/client.go]: Mean to only make a new pkScript for that channel?

File: watchtower/wtclient/session_queue.go [hunk 4]
```diff
@@ -0,0 +1,512 @@
...
+	q.queueCond.L.Lock()
+
+	// Examine the current reserve status of the session queue.
+	curStatus := q.reserveStatus()
+	switch curStatus {
+
+	// The session queue is exhausted, and cannot accept the task because it
+	// is full. Reject the task such that it can be tried against a
+	// different session.
+	case reserveExhausted:
```
[the remark this answers is not in the document]
Comment (cfromknecht) [reply watchtower/wtclient/session_queue.go]: why is that? this case should only hit if the tower has already been fully utilized

File: watchtower/wtclient/task_pipeline.go [hunk 5]
```diff
@@ -0,0 +1,184 @@
...
+		q.queueCond.L.Lock()
+		for q.queue.Front() == nil {
+			q.queueCond.Wait()
+
+			select {
+			case <-q.quit:
+				// Exit only after the queue has been fully drained.
+				if q.queue.Len() == 0 {
+					log.Debugf("Revoked state pipeline flushed.")
+					return
```
[the remark this answers is not in the document]
Comment (cfromknecht) [reply watchtower/wtclient/task_pipeline.go]: could catch, unlocks added!

File: watchtower/wtserver/server.go [hunk 6]
```diff
@@ -86,11 +88,31 @@ func New(cfg *Config) (*Server, error) {
 		cfg.ChainHash,
 	)
 
+	// Now that we've established that this session does not exist in the
```
[the remark this answers is not in the document]
Comment (cfromknecht) [reply watchtower/wtserver/server.go]: yes this is a temporary meausre until the server db has been merged, limiting it generating one addresser per restart when using an in-memory database. the server has no way to correlate sessions/updates to a particular channel, so it can't necessarily do this in the same way the client can. the final approach will likely be to keep a running count of the number of state updates allocated to a partiular address, and rotate to a new reward address after every ~N. I'm thinking this would occur on the order of 100k ~ 1M, though really depends on the relative frequency of breaches/updates-sent-to-this-tower. With some real data in the wild, it should be easier to tune this value (the tower could even self adjust in theory)

File: watchtower/wtclient/interface.go [hunk 7]
```diff
@@ -0,0 +1,53 @@
...
+	// ListActiveSessions returns all sessions that have not yet been
+	// exhausted. This is used on startup to find any sessions which may
+	// still be able to accept state updates.
+	ListActiveSessions() (map[wtdb.SessionID]*wtdb.ClientSession, error)
+}
+
+// NetDialer connects to a remote node using an authenticated transport, such as
+// brontide. The dialer argument is used to specify a resolver, which allows
+// this method to be used over Tor or clear net connections.
+type NetDialer func(localPriv *btcec.PrivateKey, netAddr *lnwire.NetAddress,
```
[the remark this answers is not in the document]
Comment (cfromknecht) [reply watchtower/wtclient/interface.go]: the issue arises in the at `brontide.Dial` is a package level method, where as the `*mockNet.AuthDial` method used in tests is a member function. In practice, we could just `brontide.Dial` within the client, though then we couldn't mock out the network at have `Dial` return a mock connection

File: watchtower/wtclient/candidate_iterator.go [hunk 8]
```diff
@@ -0,0 +1,82 @@
+package wtclient
+
+import (
+	"container/list"
+	"sync"
+
+	"github.com/lightningnetwork/lnd/watchtower/wtdb"
+)
+
+// TowerCandidateIterator provides an abstraction for iterating through possible
```
[the remark this answers is not in the document]
Comment (cfromknecht) [reply watchtower/wtclient/candidate_iterator.go]: yes my thinking here is that this could later be backed by a graph iterator in the public case. we can also expand this tower iterator to be mutable in the case that we want the set to be modified via rpc after cli configuration

File: watchtower/wtclient/client.go (hunk 1 above)
Comment (cfromknecht) [reply watchtower/wtclient/client.go]: agreed, i'm planning to do this in a follow up to prevent the scope of this pr from expanding too much

File: watchtower/wtclient/client.go (hunk 3 above)
Comment (cfromknecht) [reply watchtower/wtclient/client.go]: yes, so whenever register is called the first time we will derive and persist a new pkscript to use for that channel. if one has already been made, this will be a nop

**Comment by wpaulino** [comment]
Needs a rebase.

File: watchtower/wtclient/session_queue.go (hunk 4 above)
[the remark this answers is not in the document]
Comment (Roasbeef) [reply watchtower/wtclient/session_queue.go]: Need to update my mental model, but at the time it seemed that if the amount of slots we needed exceeds the current available number of slots in the sessions, then rather than bin-packing up the max slots and making a new session for the remainder, we'd just make an entirely new session. If this is the case, then for the historical upload case, we might loop here forever if there're more than 65k states to be backed up?

File: watchtower/wtclient/client.go (hunk 3 above)
Comment (cfromknecht) [reply watchtower/wtclient/client.go]: ended up adding this to the current working PR

File: watchtower/wtserver/server.go [hunk 9]
```diff
@@ -325,28 +347,30 @@ func (s *Server) handleCreateSession(peer Peer, id *wtdb.SessionID,
...
-	// sweep transactions destined for this tower, if its negotiated output
-	// is not dust.
-	rewardAddress, err := s.cfg.NewAddress()
-	if err != nil {
-		log.Errorf("unable to generate reward addr for %s", id)
-		return s.replyCreateSession(
-			peer, id, wtwire.CodeTemporaryFailure, nil,
-		)
-	}
+	/*
```
[the remark this answers is not in the document]
Comment (cfromknecht) [reply watchtower/wtserver/server.go]: removed

File: watchtower/wtdb/client_state_update.go [hunk 10]
```diff
@@ -0,0 +1,16 @@
+package wtdb
+
+// ClientStateUpdate holds a state update sent by a client along with its
+// SessionID.
+type ClientStateUpdate struct {
+	// ID the session id of the client who sent the state update.
+	ID SessionID
+
+	// Hint is the 16-byte prefix of the revoked commitment transaction.
```
[the remark this answers is not in the document]
Comment (cfromknecht) [reply watchtower/wtdb/client_state_update.go]: fixed

File: watchtower/wtclient/task_pipeline.go (hunk 5 above)
[the remark this answers is not in the document]
Comment (Roasbeef) [reply watchtower/wtclient/task_pipeline.go]: Unlocks still missing in current diff as I see it.

File: watchtower/wtclient/session_queue.go [hunk 11]
```diff
@@ -0,0 +1,512 @@
...
+
+	// Process the reply from the tower.
+	switch stateUpdateReply.Code {
+
+	// The tower reported a successful update, validate the response and
+	// record the last applied returned.
+	case wtwire.CodeOK:
+		lastApplied := stateUpdateReply.LastApplied
+		if lastApplied > stateUpdate.SeqNum {
+			// TODO(conner): borked watchtower?
```
Comment (Roasbeef) [inline watchtower/wtclient/session_queue.go]: At the very least, in the short term, should log this case and the one below.

File: watchtower/wtclient/session_queue.go [hunk 12]
```diff
@@ -0,0 +1,512 @@
...
+		}
+
+		if lastApplied < q.cfg.ClientSession.TowerLastApplied {
+			// TODO(conner): borked watchtower?
+		}
+
+		// TODO(conner): store last applied for session
+		// TODO(conner): delete blob locally
+		q.queueCond.L.Lock()
+		q.cfg.ClientSession.SeqNum++
```
Comment (Roasbeef) [inline watchtower/wtclient/session_queue.go]: When does this state get updated on disk?

File: watchtower/wtclient/session_queue.go [hunk 13]
```diff
@@ -0,0 +1,512 @@
...
+// NOTE: This method MUST be called with queueCond's exclusive lock held.
+func (q *sessionQueue) reserveStatus() reserveStatus {
+	seqNum := q.cfg.ClientSession.SeqNum
+	numPending := uint16(q.queue.Len())
+	maxUpdates := q.cfg.ClientSession.Policy.MaxUpdates
+
+	log.Debugf("SessionQueue %s reserveStatus seqnum=%d pending=%d "+
+		"max-updates=%d", q.ID(), seqNum, numPending, maxUpdates)
+
+	if seqNum+numPending < maxUpdates {
```
Comment (Roasbeef) [inline watchtower/wtclient/session_queue.go]: Here's the spot I'm referring to above, re underutilization: if we have 10 spots left, and we need 15 updates, then we'll make a new session to pack those 15, rather than using the remaining 10 in this session, and making a new session for the remaining 5 updates.

File: watchtower/wtclient/interface.go [hunk 14]
```diff
@@ -0,0 +1,53 @@
...
+// DB abstracts the required database operations required by the watchtower
+// client.
+type DB interface {
+	// CreateTower initialize an address record used to communicate with a
+	// watchtower. Each Tower is assigned a unique ID, that is used to
+	// amortize storage costs of the public key when used by multiple
+	// sessions.
+	CreateTower(*lnwire.NetAddress) (*wtdb.Tower, error)
+
+	MarkBackupIneligible(chanID lnwire.ChannelID, commitHeight uint64) error
```
[the remark this answers is not in the document]
Comment (cfromknecht) [reply watchtower/wtclient/interface.go]: should be gone now, my push failed last night

File: watchtower/wtclient/session_queue.go [hunk 15]
```diff
@@ -0,0 +1,512 @@
...
+			case <-q.forceQuit:
+				return
+			default:
+			}
+		}
+		q.queueCond.L.Unlock()
+
+		// Initiate a new connection to the watchtower and attempt to
+		// drain all pending tasks.
+		q.drainBackups()
```
[the remark this answers is not in the document]
Comment (cfromknecht) [reply watchtower/wtclient/session_queue.go]: since the `Dial` may block indefinitely, we first try to connect, then reacquire the mutex to pull items off. otherwise the client won't be able to continue queueing items while the dial is happening

File: watchtower/wtclient/session_negotiator.go [hunk 16]
```diff
@@ -0,0 +1,413 @@
...
+			// We've run out of addresses, double and clamp backoff.
+			backoff *= 2
+			if backoff > n.cfg.MaxBackoff {
+				backoff = n.cfg.MaxBackoff
+			}
+
+			log.Debugf("Unable to get new tower candidate, retrying "+
+				"after %v--reason: %v", backoff, err)
+
+			goto retryWithBackoff
```
[the remark this answers is not in the document]
Comment (cfromknecht) [reply watchtower/wtclient/session_negotiator.go]: `Reset` is called before reentering the `for` loop?

File: watchtower/wtclient/session_queue.go (hunk 13 above)
Comment (cfromknecht) [reply watchtower/wtclient/session_queue.go]: hmm, the latter is how this behaves in practice. maybe this will clear it up

`seqNum` is the highest allocated sequence number, and `numPending` is the number of bound tasks that are in the `pendingQueue` that have not been assigned a sequence number. since the updates are assigned sequentially (and not in batch), this method returns true up until the last allocated sequence number and number of unallocated updates is exactly `maxUpdates`

File: watchtower/wtclient/session_queue.go (hunk 12 above)
Comment (cfromknecht) [reply watchtower/wtclient/session_queue.go]: latest version properly updates the db, you're correct that the first version did this assignment in memory ;)

File: watchtower/wtclient/session_queue.go (hunk 13 above)
Comment (Roasbeef) [reply watchtower/wtclient/session_queue.go]: Ahh ok, thanks for the explanation, was under the impression that they could be dispatched in batch.

File: watchtower/wtserver/server.go [hunk 17]
```diff
@@ -55,6 +55,8 @@ type Config struct {
 
 	// ChainHash identifies the network that the server is watching.
 	ChainHash chainhash.Hash
+
+	NoAckUpdates bool
```
Comment (wpaulino) [inline watchtower/wtserver/server.go]: Why would we not want to ACK updates?

File: watchtower/wtclient/client.go [hunk 18]
```diff
@@ -0,0 +1,810 @@
...
+	// Select any candidate session at random, and remove it from the set of
+	// candidate sessions.
+	var candidateSession *wtdb.ClientSession
+	for id, sessionInfo := range c.candidateSessions {
+		delete(c.candidateSessions, id)
+
+		// Skip any sessions with policies that don't match the current
+		// configuration. These can be used again if the client changes
+		// their configuration back.
+		if sessionInfo.Policy != c.cfg.Policy {
```
Comment (wpaulino) [inline watchtower/wtclient/client.go]: What ends up happening with these sessions on the server side? Are they no longer being watched?

File: watchtower/wtclient/client.go [hunk 19]
```diff
@@ -0,0 +1,810 @@
...
+
+	// The sessionQueue has available capacity but the task was rejected,
+	// this indicates that the task was ineligible for backup.
+	case reserveAvailable:
+		c.stats.taskIneligible()
+
+		log.Infof("Backup chanid=%s commit-height=%d is ineligible",
+			task.id.ChanID, task.id.CommitHeight)
+
+		err := c.cfg.DB.MarkBackupIneligible(
```
Comment (wpaulino) [inline watchtower/wtclient/client.go]: If this is a dust error, should we perhaps try with another tower? I see that the client only has access to one tower so far, so can add a TODO if so.

File: watchtower/wtclient/session_negotiator.go [hunk 20]
```diff
@@ -0,0 +1,453 @@
...
+
+	case wtwire.CreateSessionCodeRejectRejectMaxUpdates:
+		return fmt.Errorf("tower rejected max updates: %v",
+			policy.MaxUpdates)
+
+	case wtwire.CreateSessionCodeRejectRewardRate:
+		// The tower rejected the session because of the reward rate. If
+		// we didn't request a reward session, we'll treat this as a
+		// permanent tower failure.
+		if !policy.BlobType.Has(blob.FlagReward) {
```
[the remark this answers is not in the document]
Comment (Roasbeef) [reply watchtower/wtclient/session_negotiator.go]: Possible that it could, but atm a config change like that would require a restart. In `CheckRemoteInit` we'd ensure that the server supported w/e features we deem necessary. If they say we have a bad reward rate, and the we didn't specify one at all, then it would likely point to a bug in the tower server.

File: watchtower/wtclient/client_test.go [hunk 21]
```diff
@@ -0,0 +1,1115 @@
...
+				SweepFeeRate: 1,
+			},
+		},
+		fn: func(h *testHarness) {
+			const (
+				numUpdates = 5
+				chanID     = 0
+			)
+
+			// Generate numUpdates retributions and back them pu.
```
Comment (Roasbeef) [inline watchtower/wtclient/client_test.go]: incomplete sentence it looks like

File: watchtower/wtserver/server.go (hunk 17 above)
Comment (Roasbeef) [reply watchtower/wtserver/server.go]: Useful for writing a unit test to simulate a faulty server.

File: watchtower/wtclient/session_negotiator.go [hunk 22]
```diff
@@ -0,0 +1,453 @@
...
+		case <-n.quit:
+			return ErrNegotiatorExiting
+		}
+
+	// TODO(conner): handle error codes properly
+	case wtwire.CreateSessionCodeRejectBlobType:
+		return fmt.Errorf("tower rejected blob type: %v",
+			policy.BlobType)
+
+	case wtwire.CreateSessionCodeRejectRejectMaxUpdates:
```
[the remark this answers is not in the document]
Comment (cfromknecht) [reply watchtower/wtclient/session_negotiator.go]: Good catch 😅

File: watchtower/wtclient/task_pipeline.go [hunk 23]
```diff
@@ -0,0 +1,185 @@
...
+// from accepting any further tasks, and will cause the pipeline to exit after
+// all updates have been delivered to the downstream receiver. If this process
+// hangs and is unable to make progress, users can optionally call ForceQuit to
+// abandon the reliable draining of the queue in order to permit shutdown.
+type taskPipeline struct {
+	started sync.Once
+	stopped sync.Once
+	forced  sync.Once
+
+	queueMtx  sync.Mutex
```
[the remark this answers is not in the document]
Comment (cfromknecht) [reply watchtower/wtclient/task_pipeline.go]: it has to be allocated somewhere, i would say a member field is better than allocating a mutex arbitrarily on the heap?

File: watchtower/wtclient/session_queue.go [hunk 24]
```diff
@@ -0,0 +1,690 @@
...
+	})
+
+	// Finally, add the sorted, committed updates to he commitQueue. These
+	// updates will be prioritized before any new tasks are assigned to the
+	// sessionQueue. The queue will begin uploading any tasks in the
+	// commitQueue as soon as it is started, e.g. during client
+	// initialization when detecting that this session has unacked updates.
+	for _, update := range sortedUpdates {
+		q.commitQueue.PushBack(update)
+	}
```
[the remark this answers is not in the document]
Comment (cfromknecht) [reply watchtower/wtclient/session_queue.go]: if the update is committed, then the persisted session queue should already have its sequence number incremented. the updated value is read on startup and `q.seqNum` is set to that value. we could add a sanity check to assert this if we feel its necessary

File: watchtower/wtclient/session_queue.go [hunk 25]
```diff
@@ -0,0 +1,690 @@
...
+	commitQueue  *list.List
+	pendingQueue *list.List
+	queueMtx     sync.Mutex
+	queueCond    *sync.Cond
+
+	localInit *wtwire.Init
+	towerAddr *lnwire.NetAddress
+
+	seqNum      uint16
+	lastApplied uint16
```
[the remark this answers is not in the document]
Comment (cfromknecht) [reply watchtower/wtclient/session_queue.go]: you're right, we now exclusively read this value from the db, removed

File: watchtower/wtclient/client.go [hunk 26]
```diff
@@ -0,0 +1,810 @@
...
+	// NetDialer establishes a brontide connection over an onion or clear
+	// network.
+	NetDial NetDialer
+
+	// DB provides access to the client's stable storage medium.
+	DB DB
+
+	// Net specifies the type of network the client will use to initiate
+	// outbound requests, which can be either onion or clear net.
+	Net tor.Net
```
[the remark this answers is not in the document]
Comment (cfromknecht) [reply watchtower/wtclient/client.go]: great suggestion, must cleaner

File: watchtower/wtclient/client.go [hunk 27]
```diff
@@ -0,0 +1,810 @@
...
+			case <-c.statTicker.C:
+				log.Infof("Client stats: %s", c.stats)
+			}
+
+		// No active session queue but have additional sessions.
+		case c.sessionQueue == nil && len(c.candidateSessions) > 0:
+			// We've exhausted the prior session, we'll pop another
+			// from the remaining sessions and continue processing
+			// backup tasks.
+			c.sessionQueue = c.nextSessionQueue()
```
[the remark this answers is not in the document]
Comment (cfromknecht) [reply watchtower/wtclient/client.go]: nice catch, fixed!

File: watchtower/wtmock/client_db.go [hunk 28]
```diff
@@ -0,0 +1,223 @@
...
+
+	"github.com/lightningnetwork/lnd/lnwire"
+	"github.com/lightningnetwork/lnd/watchtower/wtdb"
+)
+
+type towerPK [33]byte
+
+// ClientDB is a mock, in-memory database or testing the watchtower client
+// behavior.
+type ClientDB struct {
```
[the remark this answers is not in the document]
Comment (cfromknecht) [reply watchtower/wtmock/client_db.go]: this creates an import cycle due to the `wtclient/backup_task_internal_test.go` also importing wtmock. maybe fix in a follow up?

File: watchtower/wtserver/server_test.go [hunk 29]
```diff
@@ -448,6 +448,38 @@ var stateUpdateTests = []stateUpdateTestCase{
...
+			testnetChainHash,
+		),
+		createMsg: &wtwire.CreateSession{
+			BlobType:     blob.TypeDefault,
+			MaxUpdates:   4,
+			RewardBase:   0,
+			RewardRate:   0,
+			SweepFeeRate: 1,
+		},
+		updates: []*wtwire.StateUpdate{
```
[the remark this answers is not in the document]
Comment (cfromknecht) [reply watchtower/wtserver/server_test.go]: indeed we could, tho they test slightly different things, the former tests sending seqnum+1 after disconnect and the latter tests resending seqnum. i would prefer to keep them separate as we will likely expand on this when we increase the window to more than 1

File: watchtower/wtclient/client.go (hunk 18 above)
Comment (cfromknecht) [reply watchtower/wtclient/client.go]: the server is always watching all updates it has ever received. this check just prevents against the case where we used half of a session and restart with different parameters. we skip the old policies so that if we ever restart with the prior params, we can resume using that session

File: watchtower/wtclient/client.go (hunk 19 above)
Comment (cfromknecht) [reply watchtower/wtclient/client.go]: atm the client only accepts a single policy, so if it produces dust it will also produce dust for any tower that also accepts that policy

**Comment by cfromknecht** [comment]
@Roasbeef @wpaulino comments addressed, ptal

File: watchtower/wtmock/client_db.go (hunk 28 above)
[the remark this answers is not in the document]
Comment (wpaulino) [reply watchtower/wtmock/client_db.go]: Yeah definitely not a blocker.
